src/types/closure.md

```diff
@@ -42,48 +50,153 @@ so that the call to `f` works as if it were:
+
+`ImmBorrow < UniqueImmBorrow < MutBorrow < ByValue`.
 
-Composite types such as structs, tuples, and enums are always captured entirely,
-not by individual fields. It may be necessary to borrow into a local variable in
-order to capture a single field:
+Note that this might need to be applied recursively.
 
 ```rust
-# use std::collections::HashSet;
-#
-struct SetVec {
-    set: HashSet<u32>,
-    vec: Vec<u32>
-}
+# fn move_value<T>(_: T){}
+let s = String::from("S");
+let t = (s, String::from("T"));
+let mut u = (t, String::from("U"));
+
+let c = || {
+    println!("{:?}", u); // u captured by ImmBorrow
+    u.1.truncate(0); // u.0 captured by MutBorrow
+    move_value(u.0.0); // u.0.0 captured by ByValue
+};
+```
 
-impl SetVec {
-    fn populate(&mut self) {
-        let vec = &mut self.vec;
-        self.set.iter().for_each(|&n| {
-            vec.push(n);
-        })
-    }
-}
+Overall the closure will capture `u` by `ByValue`.
+
+### Wild Card Patterns
+Closures only capture data that needs to be read, which means the following closures will not capture `x`
+
+```rust
+let x = 10;
+let c = || {
+    let _ = x;
+};
+
+let c = || match x {
+    _ => println!("Hello World!")
+};
 ```
 
-If, instead, the closure were to use `self.vec` directly, then it would attempt
-to capture `self` by mutable reference. But since `self.set` is already
-borrowed to iterate over, the code would not compile.
+### Capturing references in move contexts
+
+Moving fields out of references is not allowed. As a result, in the case of move closures, when values accessed through a shared references are moved into the closure body, the compiler will truncate right before a dereference.
+
+```rust
+struct T(String, String);
+
+let mut t = T(String::from("foo"), String::from("bar"));
+let t = &mut t;
+let c = move || t.0.truncate(0); // closure captures `t`
+```
+
+### Raw pointer dereference
+Because it is `unsafe` to dereference a raw pointer, closures will only capture the prefix of a path that runs up to, but not including, the first dereference of a raw pointer.
+
+```rust,
+struct T(String, String);
+
+let t = T(String::from("foo"), String::from("bar"));
+let t = &t as *const T;
+
+let c = || unsafe {
+    println!("{}", (*t).0); // closure captures t
+};
+```
+
+### Reference into unaligned `struct`s
+
+Because it is `unsafe` to hold references to unaligned fields in a structure, closures will only capture the prefix of the path that runs up to, but not including, the first field access into an unaligned structure.
+
+```rust
+#[repr(packed)]
+struct T(String, String);
+
+let t = T(String::from("foo"), String::from("bar"));
+let c = || unsafe {
+    println!("{}", t.0); // closure captures t
+};
+```
```

Comment:
Yes the example is wrong and doesn't compile, but I'm not sure if your example is relevant here. I guess the `addr_of!` macro will always capture the whole item, since it is expanded into something like `packed as usize + some_offset`.

I think this example would work:
```Rust
#[repr(packed)]
struct T(i32, i32);

fn main() {
    let t = T(2, 5);
    let c = || {
        t.0; // closure captures t.0 if not packed, but t if packed
        t.1;
    };
    println!("{}", core::mem::size_of_val(&c))
}
```
Normally, `t.0` should be captured by immutable borrow (since `i32` is copy) but now that `t` is packed and taking `&t.0` is UB, the whole `t` will be captured and this code will print `8` instead of `16` when `T` is not packed.